File: app/src/main/java/com/jetpack/xhb/okio/SegmentPool.java
package com.jetpack.xhb.okio;

public class SegmentPool {

    static final long MAX_SIZE = 64 * 1024;

    static Segment next;

    static long byteCount;

    private SegmentPool() {}

    static Segment take() {
        synchronized (SegmentPool.class) {
            if (next != null) {
                Segment result = next;
                next = result.next;
                result.next = null;
                byteCount -= Segment.SIZE;
                return result;
            }
            return new Segment();
        }
    }

    static void recycle(Segment segment) {
        if (segment.next != null || segment.prev != null) {
            throw  new IllegalArgumentException();
        }

        if (segment.shared) return;

        synchronized (SegmentPool.class) {
            if (byteCount + Segment.SIZE > MAX_SIZE) return;

            byteCount += Segment.SIZE;

            segment.next = next;
            segment.pos = segment.limit = 0;
            next = segment;
        }
    }
}

File: app/src/main/java/com/jetpack/xhb/okio/kotlin/SegmentPool.kt
package com.jetpack.xhb.okio.kotlin

/**
 * object 关键字可以表达两种含义：一种是对象表达式,另一种是 对象声明
 *
 * 对象声明
 *      用object 修饰的类为静态类，里面的方法和变量都为静态的
 *      Java调用为  SegmentPool.INSTANCE.take();  （注意使用INSTANCE）
 */
object SegmentPool {

    const val MAX_SIZE = 64 * 1024L

    @JvmField var next: Segment? = null

    @JvmField var byteCount = 0L

    /**
     * let 操作符
     * let 函数(处理可空表达式 替换断语句)
     *
     * 1. let 表达式，和 ‘ ?. ’运算符一起使用，很方便
     * 如果 next != null 在lambda中it 或者自己定义的（result） 非空
     * 如果 next == null 什么都不会发生
     */
    @JvmStatic fun take(): Segment {
        synchronized(this) {
            next?.let {
                result ->
                next = result.next
                result.next = null
                byteCount -= Segment.SIZE
                return result
            }
        }
        return Segment()
    }

}